File: langchain-nextjs-template/lib/vector/qdrant-service.ts
/**
 * Qdrant Vector Database Service
 * 
 * Handles all vector operations including embeddings storage,
 * similarity search, and collection management
 * Follows SOLID principles and clean architecture patterns
 */

import { QdrantClient } from '@qdrant/js-client-rest';

// Types for better type safety
export interface VectorDocument {
    id: string;
    vector: number[];
    payload: {
        content: string;
        title?: string;
        type: 'session' | 'document' | 'knowledge' | 'file';
        userId?: string;
        sessionId?: string;
        category?: string;
        tags?: string[];
        metadata?: Record<string, any>;
        createdAt: string;
        updatedAt: string;
    };
}

export interface SearchResult {
    id: string;
    score: number;
    payload: VectorDocument['payload'];
}

export interface SearchParams {
    vector: number[];
    limit?: number;
    scoreThreshold?: number;
    filter?: Record<string, any>;
    withPayload?: boolean;
    withVector?: boolean;
}

export interface CollectionInfo {
    name: string;
    vectorsCount: number;
    indexedVectorsCount: number;
    pointsCount: number;
    segmentsCount: number;
    status: string;
    optimizerStatus: string;
    payloadSchema: Record<string, any>;
}

export class QdrantService {
    private client: QdrantClient;
    private collections: {
        sessions: string;
        documents: string;
        knowledge: string;
        files: string;
    };

    constructor(url: string = 'http://localhost:6333') {
        this.client = new QdrantClient({
            url,
            checkCompatibility: false // Disable version compatibility check
        });
        this.collections = {
            sessions: 'sessions',
            documents: 'documents',
            knowledge: 'knowledge',
            files: 'files',
        };
    }

    /**
     * Initialize all collections with proper configuration
     */
    async initializeCollections(): Promise<void> {
        try {
            const collections = Object.values(this.collections);

            for (const collectionName of collections) {
                await this.createCollectionIfNotExists(collectionName);
            }

            console.log('✅ Qdrant collections initialized successfully');
        } catch (error) {
            throw new Error(`Failed to initialize Qdrant collections: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Create collection if it doesn't exist
     */
    private async createCollectionIfNotExists(collectionName: string): Promise<void> {
        try {
            let exists = false;
            try {
                const result = await this.client.collectionExists(collectionName);
                exists = result.exists || false;
            } catch (error) {
                // If collection doesn't exist, collectionExists throws an error
                exists = false;
            }

            if (!exists) {
                try {
                    await this.client.createCollection(collectionName, {
                        vectors: {
                            size: 1536, // OpenAI embedding dimension
                            distance: 'Cosine', // Best for semantic similarity
                        },
                        optimizers_config: {
                            default_segment_number: 2,
                        },
                        replication_factor: 1,
                    });
                } catch (createError: any) {
                    // If collection already exists, that's fine
                    if (createError.message?.includes('Bad Request') || createError.status === 400) {
                        console.log(`✅ Collection '${collectionName}' already exists`);
                        return;
                    }
                    throw createError;
                }

                // Create payload index for better filtering performance
                try {
                    await this.client.createPayloadIndex(collectionName, {
                        field_name: 'type',
                        field_schema: 'keyword',
                    });

                    await this.client.createPayloadIndex(collectionName, {
                        field_name: 'userId',
                        field_schema: 'keyword',
                    });

                    await this.client.createPayloadIndex(collectionName, {
                        field_name: 'sessionId',
                        field_schema: 'keyword',
                    });

                    await this.client.createPayloadIndex(collectionName, {
                        field_name: 'category',
                        field_schema: 'keyword',
                    });

                    await this.client.createPayloadIndex(collectionName, {
                        field_name: 'tags',
                        field_schema: 'keyword',
                    });
                } catch (indexError) {
                    // Indexes might already exist, continue
                    console.log(`⚠️  Some indexes for '${collectionName}' may already exist`);
                }

                console.log(`✅ Collection '${collectionName}' created with indexes`);
            } else {
                console.log(`✅ Collection '${collectionName}' already exists`);
            }
        } catch (error) {
            throw new Error(`Failed to create collection ${collectionName}: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Store session vector
     */
    async storeSessionVector(
        sessionId: string,
        vector: number[],
        payload: {
            content: string;
            title?: string;
            userId: string;
            category?: string;
            tags?: string[];
            metadata?: Record<string, any>;
        }
    ): Promise<void> {
        try {
            const document: VectorDocument = {
                id: sessionId,
                vector,
                payload: {
                    ...payload,
                    type: 'session',
                    sessionId,
                    createdAt: new Date().toISOString(),
                    updatedAt: new Date().toISOString(),
                },
            };

            await this.client.upsert(this.collections.sessions, {
                wait: true,
                points: [document],
            });
        } catch (error) {
            throw new Error(`Failed to store session vector: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Store document vector
     */
    async storeDocumentVector(
        documentId: string,
        vector: number[],
        payload: {
            content: string;
            title: string;
            userId: string;
            category?: string;
            tags?: string[];
            metadata?: Record<string, any>;
        }
    ): Promise<void> {
        try {
            const document: VectorDocument = {
                id: documentId,
                vector,
                payload: {
                    ...payload,
                    type: 'document',
                    createdAt: new Date().toISOString(),
                    updatedAt: new Date().toISOString(),
                },
            };

            await this.client.upsert(this.collections.documents, {
                wait: true,
                points: [document],
            });
        } catch (error) {
            throw new Error(`Failed to store document vector: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Store knowledge vector
     */
    async storeKnowledgeVector(
        knowledgeId: string,
        vector: number[],
        payload: {
            content: string;
            title: string;
            category: string;
            source: string;
            tags?: string[];
            trustScore?: number;
            metadata?: Record<string, any>;
        }
    ): Promise<void> {
        try {
            const document: VectorDocument = {
                id: knowledgeId,
                vector,
                payload: {
                    ...payload,
                    type: 'knowledge',
                    createdAt: new Date().toISOString(),
                    updatedAt: new Date().toISOString(),
                },
            };

            await this.client.upsert(this.collections.knowledge, {
                wait: true,
                points: [document],
            });
        } catch (error) {
            throw new Error(`Failed to store knowledge vector: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Store file vector
     */
    async storeFileVector(
        fileId: string,
        vector: number[],
        payload: {
            content: string;
            title: string;
            sessionId: string;
            userId: string;
            fileType: string;
            tags?: string[];
            metadata?: Record<string, any>;
        }
    ): Promise<void> {
        try {
            // Validate inputs
            if (!fileId || typeof fileId !== 'string') {
                throw new Error('Invalid fileId provided');
            }

            if (!Array.isArray(vector) || vector.length !== 1536) {
                throw new Error(`Invalid vector dimensions: expected 1536, got ${vector?.length || 'undefined'}`);
            }

            // Ensure all required payload fields are strings and not undefined
            const cleanPayload = {
                content: String(payload.content || '').substring(0, 1000),
                title: String(payload.title || 'Untitled').substring(0, 100),
                type: 'file',
                sessionId: String(payload.sessionId || 'unknown').substring(0, 64),
                userId: String(payload.userId || 'unknown').substring(0, 32),
                fileType: String(payload.fileType || 'unknown').substring(0, 32),
                tags: Array.isArray(payload.tags) ? payload.tags.slice(0, 5).map(tag => String(tag).substring(0, 32)) : [],
                metadata: payload.metadata || {},
                createdAt: new Date().toISOString(),
                updatedAt: new Date().toISOString(),
            };

            const point = {
                id: fileId,
                vector: vector,
                payload: cleanPayload
            };

            console.log(`Storing vector for ${fileId} with ${vector.length} dimensions`);

            await this.client.upsert(this.collections.files, {
                wait: true,
                points: [point],
            });

            console.log(`✅ Successfully stored vector in Qdrant for ${fileId}`);
        } catch (error: any) {
            console.error('Qdrant storage error details:', {
                fileId,
                vectorLength: vector?.length,
                errorMessage: error.message,
                errorStack: error.stack
            });
            throw new Error(`Failed to store file vector: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Search sessions by vector similarity
     */
    async searchSessions(
        queryVector: number[],
        userId: string,
        params: {
            limit?: number;
            scoreThreshold?: number;
            category?: string;
            tags?: string[];
        } = {}
    ): Promise<SearchResult[]> {
        try {
            const {
                limit = 10,
                scoreThreshold = 0.7,
                category,
                tags = [],
            } = params;

            const filter: Record<string, any> = {
                must: [
                    {
                        key: 'type',
                        match: { value: 'session' },
                    },
                    {
                        key: 'userId',
                        match: { value: userId },
                    },
                ],
            };

            if (category) {
                filter.must.push({
                    key: 'category',
                    match: { value: category },
                });
            }

            if (tags.length > 0) {
                filter.must.push({
                    key: 'tags',
                    match: { any: tags },
                });
            }

            const searchResult = await this.client.search(this.collections.sessions, {
                vector: queryVector,
                limit,
                score_threshold: scoreThreshold,
                filter,
                with_payload: true,
                with_vector: false,
            });

            return searchResult.map((result: any) => ({
                id: result.id as string,
                score: result.score,
                payload: result.payload as VectorDocument['payload'],
            }));
        } catch (error) {
            throw new Error(`Failed to search sessions: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Search documents by vector similarity
     */
    async searchDocuments(
        queryVector: number[],
        userId: string,
        params: {
            limit?: number;
            scoreThreshold?: number;
            category?: string;
            tags?: string[];
        } = {}
    ): Promise<SearchResult[]> {
        try {
            const {
                limit = 10,
                scoreThreshold = 0.7,
                category,
                tags = [],
            } = params;

            const filter: Record<string, any> = {
                must: [
                    {
                        key: 'type',
                        match: { value: 'document' },
                    },
                    {
                        key: 'userId',
                        match: { value: userId },
                    },
                ],
            };

            if (category) {
                filter.must.push({
                    key: 'category',
                    match: { value: category },
                });
            }

            if (tags.length > 0) {
                filter.must.push({
                    key: 'tags',
                    match: { any: tags },
                });
            }

            const searchResult = await this.client.search(this.collections.documents, {
                vector: queryVector,
                limit,
                score_threshold: scoreThreshold,
                filter,
                with_payload: true,
                with_vector: false,
            });

            return searchResult.map((result: any) => ({
                id: result.id as string,
                score: result.score,
                payload: result.payload as VectorDocument['payload'],
            }));
        } catch (error) {
            throw new Error(`Failed to search documents: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Search knowledge base by vector similarity
     */
    async searchKnowledge(
        queryVector: number[],
        params: {
            limit?: number;
            scoreThreshold?: number;
            category?: string;
            tags?: string[];
            minTrustScore?: number;
        } = {}
    ): Promise<SearchResult[]> {
        try {
            const {
                limit = 10,
                scoreThreshold = 0.75,
                category,
                tags = [],
                minTrustScore = 0.5,
            } = params;

            const filter: Record<string, any> = {
                must: [
                    {
                        key: 'type',
                        match: { value: 'knowledge' },
                    },
                ],
            };

            if (category) {
                filter.must.push({
                    key: 'category',
                    match: { value: category },
                });
            }

            if (tags.length > 0) {
                filter.must.push({
                    key: 'tags',
                    match: { any: tags },
                });
            }

            if (minTrustScore > 0) {
                filter.must.push({
                    key: 'trustScore',
                    range: {
                        gte: minTrustScore,
                    },
                });
            }

            const searchResult = await this.client.search(this.collections.knowledge, {
                vector: queryVector,
                limit,
                score_threshold: scoreThreshold,
                filter,
                with_payload: true,
                with_vector: false,
            });

            return searchResult.map((result: any) => ({
                id: result.id as string,
                score: result.score,
                payload: result.payload as VectorDocument['payload'],
            }));
        } catch (error) {
            throw new Error(`Failed to search knowledge: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Search files by vector similarity
     */
    async searchFiles(
        queryVector: number[],
        sessionId: string,
        params: {
            limit?: number;
            scoreThreshold?: number;
            fileType?: string;
            tags?: string[];
        } = {}
    ): Promise<SearchResult[]> {
        try {
            const {
                limit = 10,
                scoreThreshold = 0.7,
                fileType,
                tags = [],
            } = params;

            const filter: Record<string, any> = {
                must: [
                    {
                        key: 'type',
                        match: { value: 'file' },
                    },
                    {
                        key: 'sessionId',
                        match: { value: sessionId },
                    },
                ],
            };

            if (fileType) {
                filter.must.push({
                    key: 'fileType',
                    match: { value: fileType },
                });
            }

            if (tags.length > 0) {
                filter.must.push({
                    key: 'tags',
                    match: { any: tags },
                });
            }

            const searchResult = await this.client.search(this.collections.files, {
                vector: queryVector,
                limit,
                score_threshold: scoreThreshold,
                filter,
                with_payload: true,
                with_vector: false,
            });

            return searchResult.map((result: any) => ({
                id: result.id as string,
                score: result.score,
                payload: result.payload as VectorDocument['payload'],
            }));
        } catch (error) {
            throw new Error(`Failed to search files: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Delete vector by ID
     */
    async deleteVector(collection: keyof typeof this.collections, id: string): Promise<void> {
        try {
            await this.client.delete(this.collections[collection], {
                wait: true,
                points: [id],
            });
        } catch (error) {
            throw new Error(`Failed to delete vector: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Update vector payload
     */
    async updateVectorPayload(
        collection: keyof typeof this.collections,
        id: string,
        payload: Record<string, any>
    ): Promise<void> {
        try {
            await this.client.setPayload(this.collections[collection], {
                wait: true,
                payload,
                points: [id],
            });
        } catch (error) {
            throw new Error(`Failed to update vector payload: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Get collection statistics
     */
    async getCollectionInfo(collection: keyof typeof this.collections): Promise<CollectionInfo> {
        try {
            const info = await this.client.getCollection(this.collections[collection]);
            return {
                name: this.collections[collection],
                vectorsCount: info.vectors_count || 0,
                indexedVectorsCount: info.indexed_vectors_count || 0,
                pointsCount: info.points_count || 0,
                segmentsCount: info.segments_count,
                status: info.status,
                optimizerStatus: typeof info.optimizer_status === 'string' ? info.optimizer_status : 'ok',
                payloadSchema: info.payload_schema,
            };
        } catch (error) {
            throw new Error(`Failed to get collection info: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Get all collections info
     */
    async getAllCollectionsInfo(): Promise<Record<string, CollectionInfo>> {
        try {
            const collections = Object.keys(this.collections) as Array<keyof typeof this.collections>;
            const info: Record<string, CollectionInfo> = {};

            for (const collection of collections) {
                info[collection] = await this.getCollectionInfo(collection);
            }

            return info;
        } catch (error) {
            throw new Error(`Failed to get all collections info: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }

    /**
     * Health check
     */
    async healthCheck(): Promise<boolean> {
        try {
            // Use getCollections as a health check since healthCheck method doesn't exist
            await this.client.getCollections();
            return true;
        } catch (error) {
            return false;
        }
    }

    /**
     * Clear collection (for testing/development)
     */
    async clearCollection(collection: keyof typeof this.collections): Promise<void> {
        try {
            // Delete all points in the collection
            await this.client.delete(this.collections[collection], {
                wait: true,
                points: [], // Empty array means delete all
            });
        } catch (error) {
            throw new Error(`Failed to clear collection: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }
}

// Export singleton instance
let qdrantServiceInstance: QdrantService | null = null;

export function getQdrantService(url?: string): QdrantService {
    if (!qdrantServiceInstance) {
        qdrantServiceInstance = new QdrantService(url);
    }
    return qdrantServiceInstance;
}

File: python-backend/app/services/qdrant_service.py
"""
Qdrant Vector Database Service
Handles all vector database operations for medical documents and knowledge
"""

import json
import uuid
from typing import Any, Dict, List, Optional, Tuple

import structlog
from qdrant_client import QdrantClient
from qdrant_client.http import models
from qdrant_client.http.models import Distance, VectorParams

from app.core.config import settings

logger = structlog.get_logger(__name__)


class QdrantService:
    """Service for managing Qdrant vector database operations"""
    
    def __init__(self):
        self.client = QdrantClient(url=settings.QDRANT_URL, check_compatibility=False)
        self.vector_size = settings.VECTOR_DIMENSIONS
        
        # Collection names
        self.MEDICAL_KNOWLEDGE_COLLECTION = "medical_knowledge"
        self.SESSIONS_COLLECTION = "sessions"
        self.DOCUMENTS_COLLECTION = "documents"
        self.CHUNKS_COLLECTION = "document_chunks"
    
    async def health_check(self) -> bool:
        """Check if Qdrant is healthy"""
        try:
            # Try to get collections list as a health check
            collections = self.client.get_collections()
            return True
        except Exception as e:
            logger.error("Qdrant health check failed", error=str(e))
            return False
    
    async def initialize_collections(self) -> bool:
        """Initialize all required collections"""
        try:
            collections = [
                self.MEDICAL_KNOWLEDGE_COLLECTION,
                self.SESSIONS_COLLECTION,
                self.DOCUMENTS_COLLECTION,
                self.CHUNKS_COLLECTION
            ]
            
            for collection_name in collections:
                await self._create_collection_if_not_exists(collection_name)
            
            logger.info("All Qdrant collections initialized successfully")
            return True
            
        except Exception as e:
            logger.error("Failed to initialize Qdrant collections", error=str(e))
            return False
    
    async def _create_collection_if_not_exists(self, collection_name: str) -> bool:
        """Create collection if it doesn't exist"""
        try:
            # Check if collection exists
            collections = self.client.get_collections()
            existing_names = [col.name for col in collections.collections]
            
            if collection_name not in existing_names:
                # Create collection with appropriate configuration
                self.client.create_collection(
                    collection_name=collection_name,
                    vectors_config=VectorParams(
                        size=self.vector_size,
                        distance=Distance.COSINE
                    )
                )
                logger.info(f"Created Qdrant collection: {collection_name}")
            else:
                logger.info(f"Qdrant collection already exists: {collection_name}")
            
            return True
            
        except Exception as e:
            logger.error(f"Failed to create collection {collection_name}", error=str(e))
            return False
    
    async def store_medical_knowledge(
        self,
        knowledge_id: str,
        vector: List[float],
        payload: Dict[str, Any]
    ) -> bool:
        """Store medical knowledge vector"""
        try:
            self.client.upsert(
                collection_name=self.MEDICAL_KNOWLEDGE_COLLECTION,
                points=[
                    models.PointStruct(
                        id=knowledge_id,
                        vector=vector,
                        payload=payload
                    )
                ]
            )
            
            logger.debug("Stored medical knowledge vector", knowledge_id=knowledge_id)
            return True
            
        except Exception as e:
            logger.error("Failed to store medical knowledge vector", knowledge_id=knowledge_id, error=str(e))
            return False
    
    async def search_medical_knowledge(
        self,
        query_vector: List[float],
        limit: int = 10,
        score_threshold: float = 0.7,
        filter_conditions: Optional[Dict[str, Any]] = None
    ) -> List[Dict[str, Any]]:
        """Search medical knowledge by vector similarity"""
        try:
            search_results = self.client.search(
                collection_name=self.MEDICAL_KNOWLEDGE_COLLECTION,
                query_vector=query_vector,
                limit=limit,
                score_threshold=score_threshold,
                query_filter=models.Filter(**filter_conditions) if filter_conditions else None
            )
            
            results = []
            for result in search_results:
                results.append({
                    "id": str(result.id),
                    "score": float(result.score),
                    "payload": result.payload
                })
            
            logger.debug("Medical knowledge search completed", results_count=len(results))
            return results
            
        except Exception as e:
            logger.error("Medical knowledge search failed", error=str(e))
            return []
    
    async def store_session_vector(
        self,
        session_id: str,
        vector: List[float],
        payload: Dict[str, Any]
    ) -> bool:
        """Store session vector"""
        try:
            self.client.upsert(
                collection_name=self.SESSIONS_COLLECTION,
                points=[
                    models.PointStruct(
                        id=session_id,
                        vector=vector,
                        payload={
                            **payload,
                            "sessionId": session_id,
                            "type": "session"
                        }
                    )
                ]
            )
            
            logger.debug("Stored session vector", session_id=session_id)
            return True
            
        except Exception as e:
            logger.error("Failed to store session vector", session_id=session_id, error=str(e))
            return False
    
    async def search_sessions(
        self,
        query_vector: List[float],
        user_id: str,
        options: Optional[Dict[str, Any]] = None
    ) -> List[Dict[str, Any]]:
        """Search sessions by vector similarity"""
        try:
            options = options or {}
            limit = options.get("limit", 10)
            score_threshold = options.get("scoreThreshold", 0.7)
            
            # Filter by user if specified
            filter_conditions = {"must": [{"key": "userId", "match": {"value": user_id}}]}
            
            search_results = self.client.search(
                collection_name=self.SESSIONS_COLLECTION,
                query_vector=query_vector,
                limit=limit,
                score_threshold=score_threshold,
                query_filter=models.Filter(**filter_conditions)
            )
            
            results = []
            for result in search_results:
                results.append({
                    "id": str(result.id),
                    "score": float(result.score),
                    "payload": result.payload
                })
            
            logger.debug("Session search completed", user_id=user_id, results_count=len(results))
            return results
            
        except Exception as e:
            logger.error("Session search failed", user_id=user_id, error=str(e))
            return []
    
    async def store_document_chunk(
        self,
        chunk_id: str,
        vector: List[float],
        payload: Dict[str, Any]
    ) -> bool:
        """Store document chunk vector"""
        try:
            self.client.upsert(
                collection_name=self.CHUNKS_COLLECTION,
                points=[
                    models.PointStruct(
                        id=chunk_id,
                        vector=vector,
                        payload={
                            **payload,
                            "chunkId": chunk_id,
                            "type": "chunk"
                        }
                    )
                ]
            )
            
            logger.debug("Stored document chunk vector", chunk_id=chunk_id)
            return True
            
        except Exception as e:
            logger.error("Failed to store document chunk vector", chunk_id=chunk_id, error=str(e))
            return False
    
    async def search_document_chunks(
        self,
        query_vector: List[float],
        document_id: Optional[str] = None,
        user_id: Optional[str] = None,
        limit: int = 10,
        score_threshold: float = 0.7
    ) -> List[Dict[str, Any]]:
        """Search document chunks by vector similarity"""
        try:
            filter_conditions = {"must": []}
            
            if document_id:
                filter_conditions["must"].append({"key": "documentId", "match": {"value": document_id}})
            
            if user_id:
                filter_conditions["must"].append({"key": "userId", "match": {"value": user_id}})
            
            search_results = self.client.search(
                collection_name=self.CHUNKS_COLLECTION,
                query_vector=query_vector,
                limit=limit,
                score_threshold=score_threshold,
                query_filter=models.Filter(**filter_conditions) if filter_conditions["must"] else None
            )
            
            results = []
            for result in search_results:
                results.append({
                    "id": str(result.id),
                    "score": float(result.score),
                    "payload": result.payload
                })
            
            logger.debug("Document chunk search completed", results_count=len(results))
            return results
            
        except Exception as e:
            logger.error("Document chunk search failed", error=str(e))
            return []
    
    async def get_all_collections_info(self) -> Dict[str, Dict[str, Any]]:
        """Get information about all collections"""
        try:
            collections = self.client.get_collections()
            info = {}
            
            for collection in collections.collections:
                collection_info = self.client.get_collection(collection.name)
                info[collection.name] = {
                    "pointsCount": collection_info.points_count,
                    "vectorsCount": collection_info.vectors_count or 0,
                    "indexedVectorsCount": collection_info.indexed_vectors_count or 0,
                    "status": collection_info.status.value if collection_info.status else "unknown"
                }
            
            return info
            
        except Exception as e:
            logger.error("Failed to get collections info", error=str(e))
            return {}
    
    async def delete_collection(self, collection_name: str) -> bool:
        """Delete a collection"""
        try:
            self.client.delete_collection(collection_name)
            logger.warning("Deleted Qdrant collection", collection=collection_name)
            return True
            
        except Exception as e:
            logger.error("Failed to delete collection", collection=collection_name, error=str(e))
            return False
    
    async def initialize_medical_knowledge_collection(self) -> bool:
        """Initialize medical knowledge collection"""
        return await self._create_collection_if_not_exists(self.MEDICAL_KNOWLEDGE_COLLECTION)
    
    async def initialize_sessions_collection(self) -> bool:
        """Initialize sessions collection"""
        return await self._create_collection_if_not_exists(self.SESSIONS_COLLECTION)
    
    async def initialize_documents_collection(self) -> bool:
        """Initialize documents collection"""
        return await self._create_collection_if_not_exists(self.DOCUMENTS_COLLECTION)
    
    def get_source_reliability_score(self, url: str) -> float:
        """Get source reliability score based on URL"""
        # Simple implementation - in production, use more sophisticated scoring
        reliable_domains = [
            "pubmed.ncbi.nlm.nih.gov",
            "www.ncbi.nlm.nih.gov",
            "www.nejm.org",
            "www.thelancet.com",
            "jamanetwork.com",
            "www.nature.com",
            "www.bmj.com",
            "www.who.int",
            "www.cdc.gov",
            "www.nih.gov"
        ]
        
        for domain in reliable_domains:
            if domain in url:
                return 0.9
        
        if any(tld in url for tld in [".edu", ".gov"]):
            return 0.8
        
        if any(tld in url for tld in [".org"]):
            return 0.7
        
        return 0.5
